src/broadcast/mod.rs

```diff
@@ -11,35 +11,47 @@
-//! `pi = (h, bi, si)`.
-//! * The proposer sends `Value(pi)` to node `i`. It translates to: "I am the proposer, and `pi`
-//! contains the `i`-th share of my value."
-//! * Every (correct) node that receives `Value(pi)` from the proposer sends it on to everyone else
-//! as `Echo(pi)`. An `Echo` translates to: "I have received `pi` directly from the proposer." If
-//! the proposer sends another `Value` message it is ignored.
-//! * So every node that receives at least _f + 1_ `Echo` messages with the same root hash can
-//! decode a value.
-//! * Every node that has received _N - f_ `Echo`s with the same root hash from different nodes
-//! knows that at least _N - 2 f_ _correct_ nodes have sent an `Echo` with that hash to everyone,
-//! and therefore everyone will eventually receive at least _N - f_ of them. So upon receiving
-//! _N - f_ `Echo`s, they send a `Ready(h)` to everyone. It translates to: "I know that everyone
-//! will eventually be able to decode the value with root hash `h`." Moreover, since every correct
-//! node only sends one kind of `Echo` message, there is no danger of receiving _N - f_ `Echo`s
-//! with two different root hashes.
-//! * Even without enough `Echo` messages, if a node receives _2 f + 1_ `Ready` messages, it knows
-//! that at least one _correct_ node has sent `Ready`. It therefore also knows that everyone will
-//! be able to decode eventually, and multicasts `Ready` itself.
-//! * If a node has received _2 f + 1_ `Ready`s (with matching root hash) from different nodes,
-//! it knows that at least _2 f + 1_ _correct_ nodes have sent it. Therefore, every correct node
-//! will eventually receive _2 f + 1_, and multicast it itself. Therefore, every correct node will
-//! eventually receive _2 f + 1_ `Ready`s, too. _And_ we know at this point that every correct
-//! node will eventually be able to decode (i.e. receive at least _2 f + 1_ `Echo` messages).
-//! * So a node with _2 f + 1_ `Ready`s and _N - 2 f_ `Echos` will decode and _output_ the value,
-//! knowing that every other correct node will eventually do the same.
+//! branch `b[i]`, the `i`-th chunk `s[i]` can be verified by anyone as belonging to the Merkle
+//! tree with root hash `h`. These values are "proof" number `i`: `p[i] = (h, b[i], s[i])`.
+//!
+//! * The proposer sends `Value(p[i])` to each node `i`.<br>
+//! It translates to:
+//! "I am the proposer, and `p[i]` contains the `i`-th share of my value."
+//! * A node that receives `Value(p[i])` from the proposer sends it on to everyone
+//! else as `Echo(p[i])`. (Additional `Value`s are ignored.)<br>
+//! It translates to:
+//! "I have received `p[i]` from the proposer."
+//! * A node that has received _N - f_ `Echo`s **or** _f + 1_ `Ready`s with root hash `h`, sends
+//! `Ready(h)` to everyone.<br>
+//! It translates to:
+//! "I know that everyone will eventually be able to decode the value with root hash `h`."
```

Comment:
> If a node has received `N-f` parts of the message, the `Ready(h)` is justified because due to the Reed-Solomon-Code this is the minimal amount of shares required to reconstruct it. The accompanying proof let's the node verify that all shares are from the same value.
>
> Sending a `Ready` after receiving `f+1` `Readys` is more speculative; since there can be at most `f` byzantine nodes "faking" a `Ready`, the value may not be complete yet, but the node is sure it will receive it eventually.

Reply:
I don't quite agree with that. I really think we send a `Ready` if and only if we know that everyone will eventually be able to decode (always under the assumption that ≤ _f_ are faulty, of course):
* _N - f_ parts are not the minimal amount to reconstruct: that's _N - 2 f_. Having received _N - f_ `Echo`s implies that _N - 2 f_ **correct** nodes have sent it, so **every node** will receive at least _N - 2 f_, so every node will be able to decode.
* _f + 1_ `Ready`s imply that at least one correct node has received _N - f_ `Echo`s, which is as good as if we had received them ourselves: at this point, we also know that everyone will eventually be able to decode (including ourselves). So I don't consider this speculative at all.

Reply:
I see now that I made a mistake there, as you pointed out, the Merkle tree lets us pinpoint which shares are incorrect, reducing the number of necessary fragments to reconstruct from *N - f* to *N - 2f*. I am not sure that that'll hold for any erasure code though; it might be a property of RS.

The second paragraph is poorly phrased on my part, the "not complete" in this case is meant to apply to just the current node, not the network itself.

I retract my objections, but I maintain we should improve this section =).

Reply:
If I understand correctly, we could use any erasure code with a code rate of _(N - 2 f) / N_ here.
(Note that I already rewrote most of this section. Please take a look at the new version.)